### PR TITLE
[Console] support eclipse-plugin type projects build with Tycho

### PR DESCRIPTION
**Pull request by HannesWell**
With this change the `MavenConsoleLineTracker` is enhanced to support `Eclipse Plug-in` projects build with Tycho.
As a consequence all recently added/enhanced features like the test-report and project links as well as automatic debugger attachment now wok for `Eclipse Plug-in` projects too.

### CODE REVIEW

**Comment by laeubi** [comment]
Just one note about the wording: Tycho projects are also Maven projects, there is no such thing as "pure" Maven or "eclipse plugin" maven.
The only difference might be that the _Tycho-Surefire plugin_ emits other messages as the _Maven Surefire Plugin_.

**Comment by laeubi** [comment]
@HannesWell is this planned for 1.20 release? If yes please assign a milestone.

**Comment by mickaelistria** [comment]
> is this planned for 1.20 release? If yes please assign a milestone.

I suggest we just stop planning for 1.20 or we will never deliver it. If it's good to merge, we merge right now, if not we delay to next release; but I don't think such improvement needs to be seen as a blocker for the release.

**Comment by akurtakov** [comment]
M3 is in 2 weeks and projects delivering new features are supposed to be done by that time. Thus any new features for 1.20 should be a total no-go from now on and effort being to release by end of next week.

**Comment by HannesWell** [comment]
> Just one note about the wording: Tycho projects are also Maven projects, there is no such thing as "pure" Maven or "eclipse plugin" maven. The only difference might be that the _Tycho_-Surefire plugin emits other messages as the _Maven_ Surefire Plugin. As both are using surefire under the hood, this might be more a matter of surefire versions anyways.

For this case the relevant difference between Eclipse Plug-in project build using Eclipse's Tycho Plug-ins for Maven and an ordinary Maven project (is ordinary better?) is that the latter is available in the `IMavenProjectRegistry` while the former is not. It is not present either because it has the Maven-nature not attached (which seems to happen frequently for Eclipse Plug-in Projects that only have a pom.xml to enhance the build) or it does not even have a pom.xml file at all (and uses a polyglot file during build).

I changed `pure Maven project` to `ordinary Maven project`. Or do you have a better phrase that is better?

File: RELEASE_NOTES.md
Comment (laeubi) [inline RELEASE_NOTES.md:19]: This does not sound correct. There are no 'ordinary' projects compared to tycho projects because they are both "maven" builds.

I would suggest to either remove this or explicitly name the mojos of what build output we support.

From the other changes it seems more that you try to support pomless builds here and those without a maven nature. In this case it has nothing to do with building with tycho I think but more with finding sources in non-m2e projects.
A better aproach might be to use **JDT** to dearch fir the type instead of searching for a certain project making assumptions about the artifact-id.

File: RELEASE_NOTES.md
Comment (laeubi) [inline RELEASE_NOTES.md:35]: See above comment

**Comment by HannesWell** [comment]
> @HannesWell is this planned for 1.20 release? If yes please assign a milestone.

Yes

> I suggest we just stop planning for 1.20 or we will never deliver it. If it's good to merge, we merge right now, if not we delay to next release; but I don't think such improvement needs to be seen as a blocker for the release.

The should be ready now (if the build succeeds).

**Review by laeubi — DISMISSED** [review]

File: org.eclipse.m2e.launching/META-INF/MANIFEST.MF
Comment (fbricon) [inline org.eclipse.m2e.launching/META-INF/MANIFEST.MF:22]: Please don't couple m2e launching with PDE. If you want to add PDE specific behavior contributed to m2e launching, please introduce an extension point.

Comment (laeubi) [reply org.eclipse.m2e.launching/META-INF/MANIFEST.MF:22]: Good catch! This would require to install PDE in all cases!

**Comment by laeubi** [comment]
> M3 is in 2 weeks and projects delivering new features are supposed to be done by that time. Thus any new features for 1.20 should be a total no-go from now on and effort being to release by end of next week.

Good point, lets finish[ these open PRs](https://github.com/eclipse-m2e/m2e-core/pulls?q=is%3Aopen+is%3Apr+milestone%3A1.20.0) and then enter "stabilization period" until end of next week. If issues arise, we can decide whether its a critical bug that needs be fixed or delay it to 2.0 ...

**Comment by mickaelistria** [comment]
> Good point, lets finish[ these open PRs](https://github.com/eclipse-m2e/m2e-core/pulls?q=is%3Aopen+is%3Apr+milestone%3A1.20.0) and then enter "stabilization period" until end of next week

I removed the PRs from the milestone. The plan now is to release 1.20.0 today, start 2.0 tomorrow and maybe branch some 1.20.1 is some regressions have happened in 1.20 and we don't feel comfortable shipping the fix in 2.0 only.

**Comment by laeubi** [comment]
> > Good point, lets finish[ these open PRs](https://github.com/eclipse-m2e/m2e-core/pulls?q=is%3Aopen+is%3Apr+milestone%3A1.20.0) and then enter "stabilization period" until end of next week
> 
> I removed the PRs from the milestone. The plan now is to release 1.20.0 today, start 2.0 tomorrow and maybe branch some 1.20.1 is some regressions have happened in 1.20 and we don't feel comfortable shipping the fix in 2.0 only.

I think the packaging type fix is valuable and quite advanced, given that the author has came up with a completely new solution on our request it seems fair to take it into account give he has a strong use-case for this.

**Comment by laeubi** [comment]
@HannesWell do you mind opening a dedicated issue for this one? We can then better track to solve this for 2.0 I think the MavenProjectregistry might be enhanced to at laest return a (temporary) project for projects that have a pom.xml but not a maven nature.

File: RELEASE_NOTES.md
Comment (HannesWell) [reply RELEASE_NOTES.md]: > This does not sound correct. There are no 'ordinary' projects compared to tycho projects because they are both "maven" builds.
> 
> I would suggest to either remove this or explicitly name the mojos of what build output we support.

This list was actually only intended as a list of examples respectively a reference for the (IMHO) most common use cases. All kind of plug-ins that launch other VMs are supported as long as one knows the right properties to command the plug-in to launch the VM the with corresponding debugger flags. The text `Listening for transport dt_socket at address:` is emitted by the JVM (you can see it also wenn launching JVM with corresponding flags manually) so I think this is quite generic.

But I think I will just remove the remark. As long as it works people are usually happy and I assume most wont except a difference in the first place any ways.

> 
> From the other changes it seems more that you try to support pomless builds here and those without a maven nature. In this case it has nothing to do with building with tycho I think but more with finding sources in non-m2e projects. A better aproach might be to use **JDT** to dearch fir the type instead of searching for a certain project making assumptions about the artifact-id.

The root problem is to find out which project from the workspace the print-out refers to. And since Maven runs in a separate process the information is not available in the Eclipse VM.
The cleanest way would probably to add an Event-Spy via a Maven-Extension to the Maven-VM that connects to the Eclipse VM and reports the information need here to the Eclipse-VM, namely groupId, artifactId and version as well as the project location. The latter can be used to search an `IProject` by its URI using the ResourcePlug-ins workspace. The ConsoleLineTracker then just has to place the links at the right text.

Igor Fedorenko has done something similar for the MavenDev view:
https://github.com/ifedorenko/com.ifedorenko.m2e.mavendev/blob/master/com.ifedorenko.m2e.mavendev.launch.ui.eventspy/src/main/java/com/ifedorenko/m2e/mavendev/launch/ui/eventspy/internal/ExecutionEventListener.java

But this would take a bit more time and I won't be able to complete this within the few next days.

File: org.eclipse.m2e.launching/META-INF/MANIFEST.MF
Comment (HannesWell) [reply org.eclipse.m2e.launching/META-INF/MANIFEST.MF:22]: With the approach suggested above coupling to PDE should not be necessary anymore.

**Comment by HannesWell** [comment]
> I removed the PRs from the milestone. The plan now is to release 1.20.0 today, start 2.0 tomorrow and maybe branch some 1.20.1 is some regressions have happened in 1.20 and we don't feel comfortable shipping the fix in 2.0 only.

OK, then lets postpone this PR.

> @HannesWell do you mind opening a dedicated issue for this one? We can then better track to solve this for 2.0 I think the MavenProjectregistry might be enhanced to at laest return a (temporary) project for projects that have a pom.xml but not a maven nature.

I can open another issue. The approach written above, will likely be a better solution because it avoids further workarounds.

**Comment by HannesWell** [comment]
I just implemented the solution described in https://github.com/eclipse-m2e/m2e-core/pull/572#discussion_r801392629, were a `EventSpy` is added that participates in the Maven build and sends the relevant Maven-project data via a UNIX domain socket to M2E running in the Eclipse-JVM (and yes that works on Windows as well. In fact I have only tested it on Windows yet).


With the possibility to consider Maven-extensions and therefore loading polyglot-projects Tycho projects are already recognize by the `MavenConsoleLineTracker`. Nevertheless the now presented approach, to obtain the 'real' data from the Maven-model in the Maven build process, the has several advantages that IMHO justify the increased complexity:
- Recognizes all projects in the workspace, even if a project has no Maven-Nature (which is the case for many PDE-projects).
- Finds the right project even if two projects with the same GAV (but different project-name) exist in the workspace.
- Avoids the need to load the Maven-model when looking up the project for a given GAV
  - Improved performance due to less file-operations while tracking the console-text
  - No need resource-locks/workspace-rules while tracking the console-text

**Comment by HannesWell** [comment]
Does anybody have any remarks or objections about the new approach?
If there are no general objections, I will continue and polish this to make it ready for submission.
Of course suggestions for improvements are welcome :)

**Comment by HannesWell** [comment]
Actually the `M2EMavenBuildProjectSpy` could be moved to the m2e.maven.runtime bundle.
This would avoid to create an extra Plugin for it, would make it more clear what dependencies are available (i.e. only what's in the Maven-runtime) and thus would prevent adding more dependencies accidentally. Furthermore it would allow to place the code for de- and encoding of the transferred data in one class and use it on the Maven and IDE site without the need to use `Eclipse-ExtensibleAPI`.

**Comment by HannesWell** [comment]
> Actually the `M2EMavenBuildProjectSpy` could be moved to the m2e.maven.runtime bundle.

This causes some difficulties during build due to the up-front manifest generation. Therefore that part should be postponed until #466 is completed. But since this is all internal it can be done at any time.

**Comment by HannesWell** [comment]
After thinking about it again, I noticed that when using Java-17 features respectively having classes compiled for Java-17 in an extension for the maven-runtime that is then always added by m2e, this effectively mandates to launch the embedded Maven-runtime with Java-17, which is artificial since Maven currently only requires Java-7 (IIRC). Therefore I have to change this back to a traditional lookback-socket connection.

**Comment by HannesWell** [comment]
For me this is ready for submission.

@fbricon, @laeubi do you want to review it again?

**Comment by laeubi** [comment]
If you are fine with it I'm also fine :-)

**Comment by HannesWell** [comment]
> If you are fine with it I'm also fine :-)

Great. Then lets see if this works out in the wild.

**Pull request merged**
